src/renderer/contexts/ExecutionContext.tsx

```diff
@@ -326,8 +332,34 @@ export const ExecutionProvider = memo(({ children }: React.PropsWithChildren<{}>
         }
     }, [backend, sendAlert]);
 
+    const statusRef = useRef(status);
+
+    const KILL_TIMEOUT = 2500;
+    useEffect(() => {
```

Comment:
I don't like this `useEffect`. Not only do we know that `useEffect` is kinda bugged in contexts, the effect might also be executed multiple times (e.g. if `restart` or `sendAlert` changes), which is not protected against. I think that this should be part of `kill` instead.

Reply:
I still need to keep the useEffect for setting the ref, but otherwise I moved everything else to be part of kill

Reply:
Why not use the original approach of `Promise.race`? This would eliminate the need for the ref.

Reply:
Because that doesn't work how you think it does. Calling `/kill` doesn't wait until it's killed. It returns immediately. So, we have to check whether or not the actual execution is over, which this timeout and status ref does.

Reply:
You're right, `/kill` doesn't work like that. I wonder if it should though. I would expect that no executor is running after `/kill` responds with "Successfully killed execution."

It wouldn't even be difficult to change `/kill`. After `ctx.executor.kill()`, just do `while ctx.executor: await sleep(1ms)` and now we wait until the executor is actually killed. Spinning is a simple solution, although it's slightly incorrect. A new `/run` request could come in after the current `/run` as killed but before the `await sleep(1ms)` yields back. That's not really a concern right now though.